### PR TITLE
Support for wait while PV reattach

### PR DESCRIPTION
**Pull request by prashanth26**
Basic working for support for reattaching works with VolumeAttachments

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #249 

**Special notes for your reviewer**:

**Release note**:

```feature operator
Draining of pods with PVs (Persistent Volume) now waits for re-attachment of PV on a different node when `volumeAttachments` support is enabled on the cluster. Else it falls back to the default PV reattachment timeout value configured. The default value is `90s` and this can be overwritten via the `machine-pv-reattach-timeout` flag. 
```

### CODE REVIEW

File: pkg/util/provider/drain/drain.go [hunk 1]
```diff
@@ -477,31 +495,40 @@ func sortPodsByPriority(pods []*corev1.Pod) {
...
-	klog.V(4).Info("PV map: ", pvMap)
 
+	// Filter the list of shared PVs
 	filterSharedPVs(pvMap)
 
-	for i := range pvMap {
-		pvList := pvMap[i]
-		vols, err := o.getVolIDsFromDriver(pvList)
+	for podHash, persistantVolumeList := range pvMap {
+		persistantVolumeListDeepCopy := persistantVolumeList
```
Comment (AxiomSamarth) [inline pkg/util/provider/drain/drain.go:515]: Did you mean to call the `DeepCopy()` method in this assignment?

**Comment by prashanth26** [comment]
The PR is ready for review. Minor things for TODOs need to be fixed. 
/invite @amshuman-kr

File: pkg/util/provider/drain/volume_attachment.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+// NewVolumeAttachmentHandler returns a new VolumeAttachmentHandler
+func NewVolumeAttachmentHandler() *VolumeAttachmentHandler {
+	return &VolumeAttachmentHandler{
+		Mutex:   sync.Mutex{},
+		workers: []chan *storagev1.VolumeAttachment{},
+	}
+}
+
+func (v *VolumeAttachmentHandler) dispatcher(obj interface{}) {
```
Comment (amshuman-kr) [inline pkg/util/provider/drain/volume_attachment.go:43]:
```suggestion
func (v *VolumeAttachmentHandler) dispatch(obj interface{}) {
```

File: pkg/util/provider/drain/volume_attachment.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	for i, worker := range v.workers {
+		klog.V(4).Infof("Dispatching request for PV %s to worker %d", *volumeAttachment.Spec.Source.PersistentVolumeName, i)
+		worker <- volumeAttachment
+	}
+}
+
+// AddVolumeAttachment is the event handler for VolumeAttachment add
+func (v *VolumeAttachmentHandler) AddVolumeAttachment(obj interface{}) {
+	klog.V(5).Infof("Adding volume attachment object")
+	v.dispatcher(obj)
```
Comment (amshuman-kr) [inline pkg/util/provider/drain/volume_attachment.go:68]:
```suggestion
	v.dispatch(obj)
```

File: pkg/util/provider/drain/volume_attachment.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// UpdateVolumeAttachment is the event handler for VolumeAttachment update
+func (v *VolumeAttachmentHandler) UpdateVolumeAttachment(oldObj, newObj interface{}) {
+	klog.V(5).Info("Updating volume attachment object")
+	v.dispatcher(newObj)
```
Comment (amshuman-kr) [inline pkg/util/provider/drain/volume_attachment.go:74]:
```suggestion
	v.dispatch(newObj)
```

File: pkg/util/provider/drain/volume_attachment.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// AddWorker is the method used to add a new worker
+func (v *VolumeAttachmentHandler) AddWorker() chan *storagev1.VolumeAttachment {
+	klog.V(4).Infof("Adding new worker. Current active workers %d", len(v.workers))
+
+	v.Lock()
+	defer v.Unlock()
+
+	newWorker := make(chan *storagev1.VolumeAttachment, 20)
```
Comment (amshuman-kr) [inline pkg/util/provider/drain/volume_attachment.go:84]: Can you please add a constant for the channel size?

File: pkg/util/provider/machinecontroller/controller.go [hunk 6]
```diff
@@ -169,18 +189,25 @@ func NewController(
...
 		// deleteMachineToSafety makes sure that orphan VM handler is invoked
 		DeleteFunc: controller.deleteMachineToSafety,
 	})
 
+	// Drain Controller Informers
+	if k8sutils.IsResourceSupported(targetCoreClient, volumeAttachmentGroupName, volumeAttachmentResourceName, volumeAttachmentResourceKind) {
+		controller.volumeAttachmentHandler = drain.NewVolumeAttachmentHandler()
+		volumeAttachmentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.volumeAttachmentHandler.AddVolumeAttachment,
+			UpdateFunc: controller.volumeAttachmentHandler.UpdateVolumeAttachment,
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/controller.go:207]: Wouldn't complete the functionality if we use the `DeleteFunc` for the `waitForDetach` as well? Or do you want to keep it for later cleanup?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Thanks for the changes @prashanth26! I have some comments below. The main one about the one goroutine per pod.

**Comment by prashanth26** [comment]
I have addressed most of the changes with [this](https://github.com/gardener/machine-controller-manager/pull/608/commits/d40e0425f3bd8d552b9beb0a8dd7419c514f57ef). The last big one for the [go routine per pod](https://github.com/gardener/machine-controller-manager/pull/608#discussion_r644182003), I shall try to address with another commit. Will update on this PR once those changes are ready.

**Comment by prashanth26** [comment]
@amshuman-kr - I have made the suggested changes in [the following separate commits](https://github.com/gardener/machine-controller-manager/pull/608/commits) for ease of review. PTAL.

Thanks for the in-depth review.

File: pkg/util/provider/drain/drain.go [hunk 7]
```diff
@@ -791,7 +856,70 @@ func (o *Options) waitForDetach(ctx context.Context, volumeIDs []string, nodeNam
...
+	}
+
+	klog.V(3).Infof("Waiting for following volumes to reattach: %v", podVolumeInfo.persistantVolumeList)
+
+	pvsWaitingForReattchments := make(map[string]bool)
+	if volumeAttachmentEventCh != nil {
+		for _, persistantVolumeName := range podVolumeInfo.persistantVolumeList {
+			pvsWaitingForReattchments[persistantVolumeName] = false
+		}
+	}
```
Comment (amshuman-kr) [inline pkg/util/provider/drain/drain.go:890]:
```suggestion
	var pvsWaitingForReattchments map[string]bool
	if volumeAttachmentEventCh != nil {
		pvsWaitingForReattchments = make(map[string]bool)
		for _, persistantVolumeName := range podVolumeInfo.persistantVolumeList {
			pvsWaitingForReattchments[persistantVolumeName] = false
		}
	}
```

**Review by amshuman-kr — APPROVED** [review]
Thanks for the changes @prashanth26! Only one very minor suggestion in the spirit of the code you had written to optimise the initialise the map. Why not optimise the construction of the map also 😁

**Pull request merged**